Support both scram and sasl plaintext authentication on top of release 0.43.0

arun-kbharathan-e3768 opened the pull request:
### Type of change

- Enhancement / new feature


### Description

- Added a new authentication type `SASL_SCRAM_AND_PLAIN` in operator to supports both sasl plaintext and scram-sha-512 login to kafka
- Kafka-init is edited to load username and passwords from secret name given in `FWSS_SECRET_NAME` and create a jaas.conf to later pass as `KAFKA_OPTS` to broker.

### Checklist

Made sure all old tests pass
Tested in EKS k8s environment.

arun-kbharathan-e3768 committed:
sasl plain text added

arun-kbharathan-e3768 committed:
push env var to init container

arun-kbharathan-e3768 committed:
test to verify sasl plain text

arun-kbharathan-e3768 committed:
sasl plain config

arun-kbharathan-e3768 committed:
init container reads secrets and geneates jaas conf

arun-kbharathan-e3768 committed:
sample for testing scram and sasl plaintext

arun-kbharathan-e3768 committed:
This permission is used while in scram_and_sasl used together to create jaas config before broker starts

arun-kbharathan-e3768 committed:
get only with label

arun-kbharathan-e3768 committed:
add permission to list get secrets

arun-kbharathan-e3768 committed:
test yamls

arun-kbharathan-e3768 committed:
set replica to 3

arun-kbharathan-e3768 committed:
delete comment

arun-kbharathan-e3768 committed:
kind and registry

arun-kbharathan-e3768 committed:
secret list get helm fix

arun-kbharathan-e3768 committed:
Update Main.java

arun-kbharathan-e3768 committed:
Update kafka.yaml

arun-kbharathan-e3768 committed:
Update InitWriter.java

arun-kbharathan-e3768 committed:
Update KafkaCluster.java

arun-kbharathan-e3768 commented on kind_registry.sh:
```diff
@@ -0,0 +1,69 @@
+#!/bin/sh
```
this i will remove
this was used to setup local k8s and docker registry using kind

arun-kbharathan-e3768 replied:
Removed

dinesh-murugiah commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ListenersUtils.java:
```diff
@@ -28,6 +29,18 @@ public class ListenersUtils {
+     */
+    public static boolean hasListenerWithSaslScramAndPLain(List<GenericKafkaListener> listeners)    {
+        return listeners.stream()
+                .anyMatch(ListenersUtils::isListenerWithSaslScramAndPlain);
```
Ideally we should allow this config only for internal listeners

arun-kbharathan-e3768 replied:
Made changes so that internal listeners only is checked.

dinesh-murugiah commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBrokerConfigurationBuilder.java:
```diff
@@ -525,6 +526,12 @@ private void configureAuthentication(String listenerName, List<String> securityP
             writer.println(String.format("listener.name.%s.sasl.enabled.mechanisms=SCRAM-SHA-512", listenerNameInProperty));
             writer.println();
+        } else if (auth instanceof KafkaListenerAuthenticationSaslScramAndPlain) {
+            securityProtocol.add(String.format("%s:%s", listenerName, getSecurityProtocol(tls, true)));
```
is it possible to allow TLS when we use plain text AUTH?

arun-kbharathan-e3768 replied:
```java
    /**
     * Generates the security protocol
     *
     * @param tls  Flag whether TLS is enabled
     * @param sasl  Flag whether SASL is enabled
     *
     * @return String with the security protocol
     */
    private String getSecurityProtocol(boolean tls, boolean sasl)   {
        String a = tls ? "SSL" : "PLAINTEXT";
        return sasl ? "SASL_" + a : a;
    }
``` 

My actual intention was for above function to return `SASL_PLAINTEXT` when called.
But when i tested it looks like PLAINTEXT works over tls too, with below listener config.
I didn't do extensive tests, but the pods came up when i gave below config

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    version: 3.8.0
    replicas: 3
    template:
      initContainer:
        securityContext:
          allowPrivilegeEscalation: true
          readOnlyRootFilesystem: false
    rack:
      topologyKey: kubernetes.io/hostname
    brokerRackInitImage: localhost:5001/arunkbha/operator:latest
    listeners:
      - name: tls
        port: 9093
        type: internal
        tls: true
        authentication:
          type: sasl_scram_and_plain
    authorization:
      type: simple
      superUsers:
        - admin
    config:
      offsets.topic.replication.factor: 3
      transaction.state.log.replication.factor: 3
      transaction.state.log.min.isr: 2
      default.replication.factor: 3
      min.insync.replicas: 2
      inter.broker.protocol.version: "3.8"
    storage:
      type: ephemeral
  zookeeper:
    replicas: 3
    storage:
      type: ephemeral
  entityOperator:
    topicOperator: {}
    userOperator: {}
``` 

This is the  diff between kafka broker configs, left side is plaintext and right side is sasl_ssl.
![sasl_tls_plain](https://github.com/user-attachments/assets/860a4c3b-3436-42fd-bd8a-1d9148600337)

dinesh-murugiah commented on docker-images/kafka-based/kafka/scripts/kafka_run.sh:
```diff
@@ -116,6 +116,10 @@ export KAFKA_OPTS
 . ./set_kafka_gc_options.sh
 
+if [ -f "/tmp/kafka_server_sasl_plain_jaas.conf" ]; then
+  KAFKA_OPTS="${KAFKA_OPTS} -Djava.security.auth.login.config=/tmp/kafka_server_sasl_plain_jaas.conf"
```
recomend to do some validations on file and print errors
simple validations like size(non empty file) can be done

arun-kbharathan-e3768 replied:
I made a change to ensure that if the file exists and is not empty, it prints a message and then continues with the execution. Is this OK?

dinesh-murugiah commented on kafka-init/src/main/java/io/strimzi/kafka/init/InitWriterConfig.java:
```diff
@@ -40,6 +47,14 @@ public class InitWriterConfig {
+    /**
+     * FWSS secret prefix to filter secrets
+     */
+    public static final ConfigParameter<String> FWSS_SECRETS_PREFIX = new ConfigParameter<>("FWSS_SECRETS_PREFIX", STRING, "fwss", CONFIG_VALUES);
```
can we get this as a configurable value from a environment variable , passed through a config map

arun-kbharathan-e3768 replied:
Here I used their way of passing environment variable. For example in kafka yaml
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    version: 3.8.0
    replicas: 3
    template:
      initContainer:
        securityContext:
          allowPrivilegeEscalation: true
          readOnlyRootFilesystem: false
        env:
          - name: FWSS_SECRETS_PREFIX
            value: fwss
    rack:
      topologyKey: kubernetes.io/hostname
    brokerRackInitImage: localhost:5001/arunkbha/operator:latest
    listeners:
      - name: plain
        port: 9092
        type: internal
        tls: false
        authentication:
          type: sasl_scram_and_plain
    authorization:
      type: simple
      superUsers:
        - admin
    config:
      offsets.topic.replication.factor: 3
      transaction.state.log.replication.factor: 3
      transaction.state.log.min.isr: 2
      default.replication.factor: 3
      min.insync.replicas: 2
      inter.broker.protocol.version: "3.8"
    storage:
      type: ephemeral
  zookeeper:
    replicas: 3
    storage:
      type: ephemeral
  entityOperator:
    topicOperator: {}
    userOperator: {}
``` 
Under `initContainer` in kafka resource they allow to pass environment variable to the initContainer, so we can pass similarly our newly added variable.
Is this fine?

dinesh-murugiah commented on kafka-init/src/main/java/io/strimzi/kafka/init/Main.java:
```diff
@@ -31,6 +33,22 @@ public static void main(String[] args) {
+        if (config.getIfAuthenticationIsSaslScramAndPlain()) {
+            String namespace = config.getNamespace();
+            if (namespace == null || namespace.isEmpty()) {
+                System.exit(1);
```
print error logs

arun-kbharathan-e3768 replied:
Done

dinesh-murugiah commented on kafka-init/src/main/java/io/strimzi/kafka/init/Main.java:
```diff
@@ -31,6 +33,22 @@ public static void main(String[] args) {
+                    .list();
+            LOGGER.info("Process Secrets");
+            if (!writer.writeFwssSecretsToJaasConf(namespace, secretList)) {
+                System.exit(1);
```
print error logs

arun-kbharathan-e3768 replied:
Done

dinesh-murugiah commented on kafka-init/src/main/java/io/strimzi/kafka/init/InitWriter.java:
```diff
@@ -109,6 +112,68 @@ private String externalAddressExport(NodeAddressType type, String address) {
+        for (Secret secret : filteredSecrets) {
+            Map<String, String> data = secret.getData();
+            for (Map.Entry<String, String> entry : data.entrySet()) {
+                String key = entry.getKey();
```
validate the key and value

arun-kbharathan-e3768 replied:
Trimmed the key variable, also adds to jass.conf only if both are non-empty.

dinesh-murugiah commented on kafka-init/src/main/java/io/strimzi/kafka/init/InitWriter.java:
```diff
@@ -109,6 +112,68 @@ private String externalAddressExport(NodeAddressType type, String address) {
+        jaasConfig.setCharAt(jaasConfig.length() - 1, ';');
+        jaasConfig.append("\n};");
+
+        return write(FILE_JAAS_CONF, jaasConfig.toString());
```
should we flush the file?

arun-kbharathan-e3768 replied:
In the write function available just below, they are doing
```java
writer.checkError()
```
And `checkError` function documentation says.
`Flushes the stream if it's not closed and checks its error state.`

So I think this is fine.

dinesh-murugiah reviewed (commented):
remove unnessary files

arun-kbharathan-e3768 committed:
Check if file exits and is not empty, it it exists it prints some messages and add to kafka opts

arun-kbharathan-e3768 committed:
removed unwanted

arun-kbharathan-e3768 committed:
Added validations to user and passwords while adding to jaas.conf

arun-kbharathan-e3768 committed:
Update InitWriterConfig.java

arun-kbharathan-e3768 committed:
added error logs before system exit

arun-kbharathan-e3768 committed:
Delete kind_registry.sh

arun-kbharathan-e3768 committed:
check  only in internal listener types for sasl_plaintext.

arun-kbharathan-e3768 committed:
Update InitWriter.java

arun-kbharathan-e3768 committed:
uses k8s client to get namespace, so removing all file read of /var/run..namespace file

arun-kbharathan-e3768 committed:
added prefix of labels to look for

arun-kbharathan-e3768 committed:
uses existing central--kafka fwss secret to load the sasl_plain username and passwords

logic to convert
central--kafka
{
  "kafka_admin": "adminusername:adminpassword",
  "zookeeper_user": "zkuser:zkpassword",
  "kafka_user_1": "kafkauser1:kafkause1password",
  "kafka_user_2": "kafkauser2:kafkauser2password",
  "kafka_user_3": "kafkauser3:kafkauser3password"
}

to

KafkaServer {
  org.apache.kafka.common.security.plain.PlainLoginModule required
  username="adminusername"
  password="adminpassword"
  user_adminusername="adminpassword"
  user_kafkause1="kafkauser1password"
  user_kafkause2="kafkauser2password"
  user_kafkause3="kafkauser3password";
};

arun-kbharathan-e3768 committed:
changed the example secret format and also added a validation to check if secret file is a valid json

arun-kbharathan-e3768 committed:
Update appendix_crds.adoc

arun-kbharathan-e3768 committed:
added rolling restart notes is plain text password is updated

arun-kbharathan-e3768 committed:
fix and log
